File: theory.js
'use strict';
const $ = id => document.getElementById(id);
const COL = { delta: '#f28b82', delta_normalised: '#6fc7bc', assoc: '#c2e899', dense_topk: '#8ab4f8', fly_measured: '#c2e899', exact: '#e7eee5', asym: '#e9be75', text: '#e7eee5', muted: '#99aa9c', line: '#2b372e' };
let last = null;
function frame(id, { xlo, xhi, ylo = 0, yhi = 1, xlog = false, xlabel = '', ylabel = '', xticks = [] }) {
  const cv = $(id), w = cv.clientWidth, h = cv.clientHeight, d = devicePixelRatio || 1;
  cv.width = w * d; cv.height = h * d; const g = cv.getContext('2d'); g.scale(d, d); g.font = '11px sans-serif';
  const L = 52, B = 34, tx = v => L + ((xlog ? Math.log10(v / xlo) / Math.log10(xhi / xlo) : (v - xlo) / (xhi - xlo))) * (w - L - 18);
  const ty = v => h - B - ((v - ylo) / (yhi - ylo)) * (h - B - 14);
  g.strokeStyle = COL.line; g.fillStyle = COL.muted;
  for (let i = 0; i <= 4; i++) { const v = ylo + (yhi - ylo) * i / 4; g.beginPath(); g.moveTo(L, ty(v)); g.lineTo(w - 14, ty(v)); g.stroke(); g.textAlign = 'right'; g.fillText(v.toFixed(v < 1 ? 2 : 0), L - 6, ty(v) + 4); }
  g.textAlign = 'center'; xticks.forEach(v => g.fillText(String(v), tx(v), h - 12));
  g.fillText(xlabel, (L + w) / 2, h - 1); g.save(); g.translate(12, h / 2); g.rotate(-Math.PI / 2); g.fillText(ylabel, 0, 0); g.restore();
  return { g, tx, ty, w, h };
}
const dot = (p, x, y, c, r = 3.5) => { p.g.fillStyle = c; p.g.beginPath(); p.g.arc(p.tx(x), p.ty(y), r, 0, 7); p.g.fill(); };
const line = (p, pts, c, dash = []) => { p.g.strokeStyle = c; p.g.lineWidth = 1.8; p.g.setLineDash(dash); p.g.beginPath(); pts.forEach(([x, y], i) => i ? p.g.lineTo(p.tx(x), p.ty(y)) : p.g.moveTo(p.tx(x), p.ty(y))); p.g.stroke(); p.g.setLineDash([]); p.g.lineWidth = 1; };
const key = (p, items, x0 = 70) => items.forEach(([label, c], i) => { p.g.fillStyle = c; p.g.textAlign = 'left'; p.g.fillText('■ ' + label, x0, 22 + i * 15); });

function draw(r) {
  const pts = r.P1.points, ov = pts.map(p => p.cross_task_overlap);
  let p = frame('c1', { xlo: Math.min(...ov) - .03, xhi: Math.max(...ov) + .03, ylo: 0, yhi: 1, xlabel: 'overlap between old- and new-class codes', ylabel: 'forgetting', xticks: [.5, .6, .7, .8, .9] });
  pts.forEach(q => { dot(p, q.cross_task_overlap, q.delta_forgetting, COL.delta); dot(p, q.cross_task_overlap, q.delta_normalised_forgetting, COL.delta_normalised); });
  line(p, [[0, r.P1.associative_forgetting_mean], [1, r.P1.associative_forgetting_mean]], COL.assoc, [5, 4]);
  key(p, [['error-driven', COL.delta], ['error-driven, normalised at test', COL.delta_normalised], ['associative rule', COL.assoc]]);
  $('t1').textContent = r.verdicts.P1.text;

  p = frame('c2', { xlo: 0.004, xhi: 0.5, ylo: .5, yhi: 1, xlog: true, xlabel: 'sparsity f (fraction of units active)', ylabel: 'accuracy', xticks: [0.005, 0.01, 0.02, 0.05, 0.1, 0.2, 0.4] });
  const styles = { 'digits|dense_topk': [COL.dense_topk, [4, 3]], 'digits|fly_measured': [COL.fly_measured, [4, 3]], 'two_shape|dense_topk': [COL.dense_topk, []], 'two_shape|fly_measured': [COL.fly_measured, []] };
  for (const [k, c] of Object.entries(r.P2.curves)) { const [col, dash] = styles[k]; line(p, c.map(x => [x.sparsity, x.accuracy]), col, dash); const pk = r.P2.peaks[k]; if (k.startsWith('two_shape')) dot(p, pk.sparsity, pk.accuracy, col, 5); }
  key(p, [['two-shape classes (solid)', COL.muted], ['digits (dashed)', COL.muted], ['dense random', COL.dense_topk], ['measured PN→KC', COL.fly_measured]]);
  $('t2').textContent = r.verdicts.P2.text;

  p = frame('c3', { xlo: 0, xhi: Math.PI / 2, ylo: 0, yhi: 1, xlabel: 'angle between inputs (radians)', ylabel: 'code overlap / k', xticks: [0, 0.5, 1, 1.5] });
  for (const [w, v] of Object.entries(r.P3.wirings)) v.points.forEach(q => dot(p, q.theta, q.overlap, COL[w], 1.6));
  line(p, r.P3.theory.map(t => [t.theta, t.exact]), COL.exact);
  line(p, r.P3.theory.map(t => [t.theta, t.asymptotic]), COL.asym, [5, 4]);
  key(p, [['exact bivariate-normal curve', COL.exact], ['f^tan²(θ/2)', COL.asym], ['dense random codes', COL.dense_topk], ['measured PN→KC codes', COL.fly_measured]], p.w - 250);
  $('t3').textContent = r.verdicts.P3.text;

  p = frame('c4', { xlo: 9, xhi: 440, ylo: 0, yhi: 1, xlog: true, xlabel: 'number of classes stored', ylabel: 'accuracy', xticks: [10, 20, 50, 100, 200, 400] });
  const cols = [COL.dense_topk, COL.assoc, COL.asym];
  Object.entries(r.P4.curves).forEach(([f, v], i) => { line(p, v.points.map(q => [q.classes, q.predicted]), cols[i], [5, 4]); v.points.forEach(q => dot(p, q.classes, q.simulated, cols[i])); });
  key(p, Object.keys(r.P4.curves).map((f, i) => [`f = ${f} (dots simulated, dashed predicted)`, cols[i]]), p.w - 300);
  $('t4').textContent = r.verdicts.P4.text;

  for (const k of ['P1', 'P2', 'P3', 'P4']) { const e = $('v' + k); e.textContent = r.verdicts[k].supported ? 'supported' : 'not supported'; e.className = 'verdict ' + (r.verdicts[k].supported ? 'y' : 'n'); }
}
async function load(recompute) {
  $('thRun').disabled = true; $('status').textContent = recompute ? 'Running…' : 'Loading…'; $('status').className = 'wb-status';
  try {
    const res = recompute ? await fetch('/api/theory', { method: 'POST', headers: { 'Content-Type': 'application/json' }, body: JSON.stringify({ seed: +$('seed').value }) })
                          : await fetch('/api/theory-benchmark');
    const r = await res.json(); if (!res.ok) throw new Error(r.error || res.status);
    last = r; draw(r); $('status').textContent = `seed ${r.parameters.seed}` + (recompute ? '' : ' (saved run)');
  } catch (e) { $('status').textContent = e.message; $('status').className = 'wb-status err'; }
  finally { $('thRun').disabled = false; }
}
$('thRun').onclick = () => load(true);
window.addEventListener('resize', () => last && draw(last));
load(false);

File: theory.py
"""Tests of the mathematical model behind the sparse-expansion experiment.

Model: z(x) = top-k indicator of Wx (sparsity f = k/N); associative readout w_c = sum of class-c codes;
prediction argmax_c <z, w_c>/||w_c||. Error-driven readout: softmax cross-entropy gradient.

P1  Why the error-driven rule forgets. Two candidate causes: (a) interference, new-task updates damage old
    classes on shared active units, which predicts forgetting to grow with cross-task code overlap; and
    (b) recency bias, training only on new classes inflates their scores so they win everywhere. Test (b)
    by comparing classes with normalised weights and no bias at test time, as the associative readout does.
P2  For classes with several shapes, associative accuracy peaks at an intermediate sparsity.
P3  For a Gaussian random W, the expected code overlap of two inputs at angle theta is exactly
        E|z(x) ∩ z(x')| / k = P(Z1 > t, Z2 > t) / f,  (Z1, Z2) standard bivariate normal with correlation
        cos(theta), t = Φ^-1(1 - f); to leading order in small f this is f ** tan²(theta / 2).
P4  With C classes, the associative readout errs when some wrong class outscores the right one. Modelling
    the right-class score as N(mu_s, sd_s) and each wrong-class score as independent N(mu_o, sd_o),
        P(correct) = ∫ φ((s - mu_s)/sd_s)/sd_s · Φ((s - mu_o)/sd_o) ** (C - 1) ds.
    The score statistics are measured at C = 10 only and the curve is extrapolated to C = 400.

Everything here is a model experiment. The digit tasks are from expansion.py; the capacity test uses
synthetic classes defined directly on projection-neuron activity.
"""
import argparse
import json
import math

import numpy as np
from scipy.stats import multivariate_normal, norm

from expansion import Associative, Delta, Expansion, TASK_SETS, load_digits, split
from learning import LearningCircuit
from model import BASE

SPARSITIES = [0.005, 0.01, 0.02, 0.05, 0.1, 0.2, 0.4]
WIRINGS = {'dense_topk': 'Dense random projection', 'fly_measured': 'Measured PN→KC'}
DEFAULT = dict(seed=7)


# ---------- analytic pieces ----------
def overlap_exact(theta, f):
    """E|z ∩ z'|/k for Gaussian W: joint upper-tail probability of a bivariate normal, divided by f."""
    rho = float(np.clip(math.cos(theta), -0.999999, 0.999999))
    if rho > 0.999: return 1.0
    t = norm.ppf(1 - f)
    both = 1 - 2 * norm.cdf(t) + multivariate_normal([0, 0], [[1, rho], [rho, 1]]).cdf([t, t])
    return float(max(both, 0.0) / f)


def overlap_asymptotic(theta, f):
    return float(f ** (math.tan(theta / 2) ** 2)) if theta < math.pi else 0.0


def capacity_accuracy(C, mu_s, sd_s, mu_o, sd_o):
    """Order-statistics model: probability the right class beats all C-1 wrong classes."""
    s = np.linspace(mu_s - 6 * sd_s, mu_s + 6 * sd_s, 2001)
    integrand = norm.pdf(s, mu_s, sd_s) * norm.cdf((s - mu_o) / sd_o) ** (C - 1)
    return float(np.trapezoid(integrand, s)) if hasattr(np, 'trapezoid') else float(np.trapz(integrand, s))


class DeltaNormalised(Delta):
    """Error-driven learning, but classes are compared by normalised weights with no bias, the way the
    associative readout does. Isolates recency bias from weight interference."""
    def predict(self, z):
        n = np.linalg.norm(self.W, axis=1); n[n == 0] = 1
        return (z @ (self.W / n[:, None]).T).argmax(1)


def pearson(a, b):
    a, b = np.asarray(a, float), np.asarray(b, float)
    return float(np.corrcoef(a, b)[0, 1]) if len(a) > 2 and a.std() > 0 and b.std() > 0 else float('nan')


# ---------- simulations ----------
class Theory:
    def __init__(self, circuit=None):
        self.c = circuit or LearningCircuit(); self.x, self.y = load_digits()

    def validate(self, o):
        p = DEFAULT | o
        v = p['seed']
        if isinstance(v, bool) or not isinstance(v, (int, float)) or v != int(v) or not 0 <= v <= 999999: raise ValueError('seed must be an integer in [0,999999]')
        p['seed'] = int(v); return p

    def protocol(self, f_tr, y_tr, f_te, y_te, tasks, head, epochs, lr, rng):
        R = np.full((len(tasks), len(tasks)), np.nan)
        for i, task in enumerate(tasks):
            idx = np.flatnonzero(np.isin(y_tr, task))
            for _ in range(epochs):
                o = rng.permutation(idx)
                for s in range(0, len(o), 16): head.fit_batch(f_tr[o[s:s + 16]], y_tr[o[s:s + 16]], lr)
            pred = head.predict(f_te)
            for j in range(i + 1):
                m = np.isin(y_te, tasks[j]); R[i, j] = (pred[m] == y_te[m]).mean()
        final = float((head.predict(f_te) == y_te).mean())
        forgetting = float(np.mean([np.nanmax(R[:, j]) - R[-1, j] for j in range(len(tasks) - 1)]))
        return final, forgetting

    @staticmethod
    def cross_task_overlap(f_tr, y_tr, tasks):
        """Mean cosine between class-mean codes of classes that belong to different tasks."""
        means = {c: f_tr[y_tr == c].mean(0) for t in tasks for c in t}
        vals = []
        for i, a in enumerate(tasks):
            for b in tasks[i + 1:]:
                for ca in a:
                    for cb in b:
                        u, v = means[ca], means[cb]; vals.append(float(u @ v / (np.linalg.norm(u) * np.linalg.norm(v) + 1e-12)))
        return float(np.mean(vals))

    def codes(self, wiring, seed, f, images):
        e = Expansion(self.c, wiring, seed, f); return e, e(images)

    def p1_p2(self, seed):
        rows = []
        for task in ['digits', 'two_shape']:
            spec = TASK_SETS[task]; tasks = spec['tasks']; labels = spec['relabel'](self.y)
            tr, te = split(self.x, self.y, seed, 0)
            for wiring in WIRINGS:
                for f in SPARSITIES:
                    _, z_tr = self.codes(wiring, seed, f, self.x[tr]); _, z_te = self.codes(wiring, seed, f, self.x[te])
                    y_tr, y_te = labels[tr], labels[te]
                    acc_a, forget_a = self.protocol(z_tr, y_tr, z_te, y_te, tasks, Associative(z_tr.shape[1]), 3, 0.05, np.random.default_rng(seed + 7))
                    row = {'task': task, 'wiring': wiring, 'sparsity': f, 'associative_accuracy': acc_a, 'associative_forgetting': forget_a,
                           'cross_task_overlap': self.cross_task_overlap(z_tr, y_tr, tasks)}
                    if task == 'digits':
                        for name, head in [('delta', Delta), ('delta_normalised', DeltaNormalised)]:
                            a_, f_ = self.protocol(z_tr, y_tr, z_te, y_te, tasks, head(z_tr.shape[1]), 3, 0.5, np.random.default_rng(seed + 7))
                            row |= {f'{name}_accuracy': a_, f'{name}_forgetting': f_}
                    rows.append(row)
        d = [r for r in rows if r['task'] == 'digits']
        keys = ['wiring', 'sparsity', 'cross_task_overlap', 'delta_forgetting', 'delta_normalised_forgetting', 'delta_accuracy', 'delta_normalised_accuracy']
        p1 = {'points': [{k: r[k] for k in keys} for r in d],
              'pearson_r': pearson([r['cross_task_overlap'] for r in d], [r['delta_forgetting'] for r in d]),
              'pearson_r_normalised': pearson([r['cross_task_overlap'] for r in d], [r['delta_normalised_forgetting'] for r in d]),
              'recency_bias_share': float(np.mean([r['delta_forgetting'] - r['delta_normalised_forgetting'] for r in d])),
              'best_recency_drop': float(max(r['delta_forgetting'] - r['delta_normalised_forgetting'] for r in d)),
              'associative_forgetting_mean': float(np.mean([r['associative_forgetting'] for r in d]))}
        slope, intercept = np.polyfit([r['cross_task_overlap'] for r in d], [r['delta_forgetting'] for r in d], 1)
        p1['fit'] = {'slope': float(slope), 'intercept': float(intercept)}
        p2 = {'curves': {}, 'peaks': {}}
        for task in ['digits', 'two_shape']:
            for wiring in WIRINGS:
                pts = [(r['sparsity'], r['associative_accuracy']) for r in rows if r['task'] == task and r['wiring'] == wiring]
                p2['curves'][f'{task}|{wiring}'] = [{'sparsity': s, 'accuracy': a} for s, a in pts]
                best = max(pts, key=lambda t: t[1])
                p2['peaks'][f'{task}|{wiring}'] = {'sparsity': best[0], 'accuracy': best[1],
                                                   'interior': best[0] not in (SPARSITIES[0], SPARSITIES[-1])}
        return p1, p2, rows

    def p3(self, seed, f=0.05, pairs=600):
        rng = np.random.default_rng(seed + 11)
        out = {'sparsity': f, 'wirings': {}}
        grid = np.linspace(0.02, math.pi / 2, 40)
        out['theory'] = [{'theta': float(t), 'exact': overlap_exact(t, f), 'asymptotic': overlap_asymptotic(t, f)} for t in grid]
        # Pairs mixing same-class and different-class images, to cover a range of angles.
        a = rng.integers(0, len(self.x), pairs); b = np.where(rng.random(pairs) < 0.5, rng.integers(0, len(self.x), pairs), a)
        noise = rng.normal(0, 0.15, (pairs, 64)) * (a == b)[:, None]
        xa, xb = self.x[a], np.clip(self.x[b] + noise, 0, 1)
        for wiring in WIRINGS:
            e = Expansion(self.c, wiring, seed, f)
            pa, pb = xa[:, e.pix_to_pn], xb[:, e.pix_to_pn]
            theta = np.arccos(np.clip((pa * pb).sum(1) / (np.linalg.norm(pa, axis=1) * np.linalg.norm(pb, axis=1) + 1e-12), -1, 1))
            za, zb = e(xa), e(xb); ov = (za * zb).sum(1) / e.k_active
            exact = np.array([overlap_exact(t, f) for t in theta])
            out['wirings'][wiring] = {'points': [{'theta': float(t), 'overlap': float(o)} for t, o in zip(theta, ov)],
                                      'rmse_vs_exact': float(np.sqrt(np.mean((ov - exact) ** 2))),
                                      'mean_bias_vs_exact': float(np.mean(ov - exact))}
        return out

    def p4(self, seed, f_list=(0.01, 0.05, 0.2), Cs=(10, 20, 50, 100, 200, 400), noise=3.0, n_train=5, n_test=5):
        rng = np.random.default_rng(seed + 13)
        npn = self.c.np; N = 4064
        W = rng.normal(0, 1 / math.sqrt(npn), (N, npn))
        protos = rng.normal(0, 1, (max(Cs), npn))
        def code(v, k):
            d = v @ W.T; out = np.zeros_like(d); top = np.argpartition(-d, k - 1, axis=1)[:, :k]; np.put_along_axis(out, top, 1.0, axis=1); return out
        res = {'noise': noise, 'n_train': n_train, 'n_test': n_test, 'curves': {}}
        for f in f_list:
            k = max(1, int(round(f * N)))
            tr = protos.repeat(n_train, 0) + noise * rng.normal(0, 1, (max(Cs) * n_train, npn))
            te = protos.repeat(n_test, 0) + noise * rng.normal(0, 1, (max(Cs) * n_test, npn))
            ztr, zte = code(tr, k), code(te, k)
            w = ztr.reshape(max(Cs), n_train, N).sum(1); w /= np.linalg.norm(w, axis=1, keepdims=True)
            scores = zte @ w.T                                                   # (test, classes)
            ytest = np.arange(max(Cs)).repeat(n_test)
            # Score statistics measured with the first 10 classes only.
            m10 = ytest < 10; s10 = scores[m10][:, :10]; right = s10[np.arange(len(s10)), ytest[m10]]
            wrong = s10[np.arange(10).repeat(n_test)[:, None] != np.arange(10)[None, :]]
            stats = {'mu_s': float(right.mean()), 'sd_s': float(right.std() + 1e-9), 'mu_o': float(wrong.mean()), 'sd_o': float(wrong.std() + 1e-9)}
            # The Gaussian model assumes symmetric wrong-class scores; sparse codes make them right-skewed,
            # so rare high scores are underestimated and the prediction is optimistic.
            skew = float(((wrong - wrong.mean()) ** 3).mean() / (wrong.std() ** 3 + 1e-12))
            pts = []
            for C in Cs:
                m = ytest < C; sc = scores[m][:, :C]
                pts.append({'classes': C, 'simulated': float((sc.argmax(1) == ytest[m]).mean()),
                            'predicted': capacity_accuracy(C, **stats)})
            gap = (stats['mu_s'] - stats['mu_o']) / stats['sd_o']
            res['curves'][str(f)] = {'stats_at_10_classes': stats, 'wrong_score_skew': skew, 'points': pts,
                                     'mean_signed_error': float(np.mean([p['predicted'] - p['simulated'] for p in pts])),
                                     'mean_abs_error': float(np.mean([abs(p['simulated'] - p['predicted']) for p in pts])),
                                     'separation': float(gap)}
        return res

    def run(self, **options):
        p = self.validate(options); seed = p['seed']
        p1, p2, rows = self.p1_p2(seed)
        p3 = self.p3(seed); p4 = self.p4(seed)
        verdicts = {
            'P1': {'supported': bool(p1['recency_bias_share'] > 0.05),
                   'text': (f"Code overlap does not explain forgetting under the error-driven rule (r = {p1['pearson_r']:.2f} over "
                            f"{len(p1['points'])} sparsity-wiring points). Comparing classes by normalised weights without a bias term "
                            f"cuts forgetting by {p1['recency_bias_share']:.2f} on average and up to {p1['best_recency_drop']:.2f}, so recency bias "
                            f"dominates. The associative readout, which normalises by construction, forgets {p1['associative_forgetting_mean']:.2f}.")},
            'P2': {'supported': bool(all(v['interior'] for k, v in p2['peaks'].items() if k.startswith('two_shape'))),
                   'text': 'Two-shape classes peak at sparsity ' + ', '.join(f"{v['sparsity']:g} ({WIRINGS[k.split('|')[1]]})" for k, v in p2['peaks'].items() if k.startswith('two_shape')) + '.'},
            'P3': {'supported': bool(p3['wirings']['dense_topk']['rmse_vs_exact'] < 0.05),
                   'text': f"Dense random codes follow the exact overlap curve (RMSE {p3['wirings']['dense_topk']['rmse_vs_exact']:.3f}); measured wiring deviates (RMSE {p3['wirings']['fly_measured']['rmse_vs_exact']:.3f}, mean bias {p3['wirings']['fly_measured']['mean_bias_vs_exact']:+.3f})."},
            'P4': {'supported': bool(all(v['mean_abs_error'] < 0.08 for k, v in p4['curves'].items() if float(k) >= 0.05)),
                   'text': ('Capacity extrapolated from 10 classes to 400: ' + ', '.join(f"mean error {v['mean_abs_error']:.3f} at f = {k}" for k, v in p4['curves'].items())
                            + '. The prediction is optimistic everywhere (signed error '
                            + ', '.join(f"{v['mean_signed_error']:+.3f}" for v in p4['curves'].values())
                            + ') and worst for the sparsest codes, where wrong-class scores are right-skewed (skew '
                            + ', '.join(f"{v['wrong_score_skew']:+.2f} at f = {k}" for k, v in p4['curves'].items())
                            + ') rather than the Gaussian the model assumes.')},
        }
        return {'parameters': p, 'sparsities': SPARSITIES, 'wirings': WIRINGS, 'P1': p1, 'P2': p2, 'P3': p3, 'P4': p4,
                'sweep': rows, 'verdicts': verdicts,
                'note': 'Single-seed model tests. Verdict thresholds (recency drop > 0.05, interior peak, RMSE < 0.05, mean error < 0.08) are set in '
                        'theory.py. P1 originally predicted that forgetting would track code overlap; the simulation refuted it and the test now '
                        'compares the two candidate mechanisms instead.'}


if __name__ == '__main__':
    ap = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    ap.add_argument('--seed', type=int, default=7); a = ap.parse_args()
    r = Theory().run(seed=a.seed)
    (BASE / 'data/theory-benchmark.json').write_text(json.dumps(r, allow_nan=False))
    for k, v in r['verdicts'].items(): print(k, 'SUPPORTED' if v['supported'] else 'NOT SUPPORTED', '-', v['text'])
